services/ingestion/block_tracking_subscriber.go

```diff
@@ -0,0 +1,277 @@
+				if !ok {
+					var err error
+					err = errs.ErrDisconnected
+					if ctx.Err() != nil {
+						err = ctx.Err()
+					}
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				blockEvents, err := r.evmEventsForHeight(ctx, blockHeader.Height)
+				if err != nil {
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				evmEvents := models.NewSingleBlockEvents(blockEvents)
+				// if events contain an error, or we are in a recovery mode
+				if evmEvents.Err != nil || r.recovery {
+					evmEvents = r.recover(ctx, blockEvents, evmEvents.Err)
+					// if we are still in recovery go to the next event
+					if r.recovery {
+						continue
+					}
+				}
+
+				for _, evt := range blockEvents.Events {
+					r.keyLock.UnlockKey(evt.TransactionID)
+				}
+				r.keyLock.Notify(blockHeader.Height)
+				lastReceivedHeight = blockHeader.Height
+
+				eventsChan <- evmEvents
+
+			case err, ok := <-errChan:
+				if !ok {
+					var err error
+					err = errs.ErrDisconnected
+					if ctx.Err() != nil {
+						err = ctx.Err()
+					}
+					eventsChan <- models.NewBlockEventsError(err)
+					return
+				}
+
+				if strings.Contains(errors.Unwrap(err).Error(), "DeadlineExceeded") ||
+					strings.Contains(errors.Unwrap(err).Error(), "unexpected EOF") {
+					blockHeadersChan, errChan, err = r.client.SubscribeBlockHeadersFromStartHeight(
+						ctx,
+						lastReceivedHeight+1,
+						flow.BlockStatusFinalized,
+					)
+					if err != nil {
+						eventsChan <- models.NewBlockEventsError(
+							fmt.Errorf(
+								"failed to subscribe for finalized block headers on height: %d, with: %w",
+								height,
+								err,
+							),
+						)
+						return
+					}
+				} else {
+					eventsChan <- models.NewBlockEventsError(fmt.Errorf("%w: %w", errs.ErrDisconnected, err))
+					return
+				}
+			}
+		}
+	}()
+
+	return eventsChan
+}
+
+func (r *RPCBlockTrackingSubscriber) evmEventsForHeight(
+	ctx context.Context,
+	height uint64,
+) (flow.BlockEvents, error) {
+	eventTypes := blocksFilter(r.chain).EventTypes
+	evmBlockEvent := eventTypes[0]
+
+	evts, err := r.client.GetEventsForHeightRange(
+		ctx,
+		evmBlockEvent,
+		height,
+		height,
+	)
+	if err != nil {
+		return flow.BlockEvents{}, err
+	}
+
+	if len(evts) != 1 && len(evts[0].Events) != 1 {
+		return flow.BlockEvents{}, fmt.Errorf(
+			"received unexpected number of EVM events for height: %d, got: %d, expected: 1",
+			height,
+			len(evts),
+		)
+	}
+
+	blockEvents := evts[0]
+	payload, err := events.DecodeBlockEventPayload(blockEvents.Events[0].Value)
+	if err != nil {
+		return flow.BlockEvents{}, err
+	}
+
+	if payload.TransactionHashRoot == types.EmptyTxsHash {
+		return blockEvents, nil
+	}
+
+	evmTxEvent := eventTypes[1]
+	evts, err = r.client.GetEventsForHeightRange(
+		ctx,
+		evmTxEvent,
+		height,
+		height,
+	)
+	if err != nil {
+		return flow.BlockEvents{}, err
+	}
+
+	if len(evts) != 1 && len(evts[0].Events) != 1 {
```

Comment:
```suggestion
	if len(evts) != 1 || len(evts[0].Events) != 1 {
```

Reply:
Properly handled in https://github.com/onflow/flow-evm-gateway/pull/728/commits/bfe61889f9d71293354569b3576e0839038a2cb7.